### ISSUE TITLE
Trivial issue with cursos in the inferior mode

### ISSUE THREAD

**Issue by ledohod** [emacs-ess/ESS#1137]
Hi,

A rather trivial issue with cursor going "backward" when exiting from insert mode to normal if the line is empty. It works as expected in non-empty lines. I have this in my init file:
```
(add-hook 'inferior-ess-r-mode-hook
          (lambda () (setq-local evil-move-cursor-back nil)))
```

**Comment by jabranham** [comment]
@lionel- you use evil-mode, right? Can you reproduce?

**Comment by ledohod** [comment]
```
Emacs version: GNU Emacs 26.3 (build 2, x86_64-pc-linux-gnu, GTK+ Version 3.24.14) of 2020-03-26, modified by Debian
Operating System: Ubuntu 20.04.2 LTS (Focal Fossa)
Evil version: Evil version 1.14.0
Evil installation type: using straight
```
Yes, I use the `evil-mode`, I'll be honest not sure how to reproduce. Above are details of the installation. Other than that, I run `R`, go to the inferior mode window, go into the insert mode and type and then Escape. I noticed one thing. It doesn't seem to be about whether the line is empty or not as in my original post, but about whether there is space after the cursor. So, if there is no space then cursor goes back, but if there is space then cursor stays put as I go to the normal mode. So, if I type `ls()` and the cursor is between the brackets, after escaping it stays where it is, but if the cursor is on the last, closing bracket then escaping makes the cursor jump back. If the line is empty, I can't insert spaces anyway, so escaping makes the cursor jump back and occupy the space between ">" prompt and the "insertable area". So, if I press "i" then the cursor is right after ">". This doesn't happen in `shell`. I noticed this change in cursor behaviour after one of the updates (not sure which it was, the `evil` or the `ess`, a while ago.

**Comment by lionel-** [comment]
I have the following hooked in `inferior-ess-mode-hook`. I think this should fix the problem.

```elisp
;; Set field boundaries in comint
;; See https://stat.ethz.ch/pipermail/ess-help/2001-July/000575.html
(setq-local comint-use-prompt-regexp nil)
```

Maybe we should detect evil usage and set this if detected? Personally I think it should even be the default because it makes iESS behave and feel like a real terminal. Advanced users who want full editability of the process buffer can set `comint-use-prompt-regexp` to `t`.

**Comment by ledohod** [comment]
Yep, that fixed it! Thank you so much! Although, it is just aesthetics, it was still a tad annoying :).

**Comment by lionel-** [comment]
It's not just aesthetics, the default behaviour gets in the way of all linewise operations such as replace-line or copy-line.
